File: app/actors/Protocol.scala
package actors

object Protocol {
  sealed trait Message

  case object CreateRoom extends Message

  case class JoinRoom(room: String, user: String) extends Message

  case object LeaveRoom extends Message

  case class ChangeReady(name: String, ready: Boolean) extends Message

  case object UnreadyAll extends Message

  case class RoomStatus(title: String, users: Map[String, Boolean]) extends Message

  case object RequestStatus extends Message
}

File: server/app/assets/js/protocol.js
var Protocol = function(ws) {
  var send = function(json) {
    ws.send(JSON.stringify(json));
  };

  var buildMsg = function($type, data) {
    var msg = { $type: "actors.Protocol." + $type };
    for (k in data) msg[k] = data[k];
    return msg;
  }

  return {
    parse: function(string) {
      var obj = JSON.parse(string);
      obj.$type = obj.$type.substring(obj.$type.lastIndexOf(".") + 1);
      return obj;
    },
    send: send,
    requestStatus: function() {
      send(buildMsg("RequestStatus"));
    },
    updateReady: function(name, newStatus) {
      send(buildMsg("ChangeReady", { name: name, ready: newStatus }));
    },
    unreadyAll: function() {
      send(buildMsg("UnreadyAll"));
    },
    kickUser: function(name) {
      send(buildMsg("KickUser", { name: name }));
    },
    changeRoomLock: function(room, locked) {
      send(buildMsg("ChangeRoomLock", { room: room, locked: locked }));
    }
  };
};
